### DISCUSSION TITLE
DateUtil.convertDate: Problems with ISO Timezones and Milliseconds

### DISCUSSION THREAD

**Discussion by NoraTheExplorer** [nextgenhealthcare/connect#5808] [Q&A] [answered]
Hi,

I'm currently expiriencing issues with converting ISO Time to HL7 timestamps if including milliseconds.

So this is my timestamp in ISO: 2023-06-01T09:51:39.9470000Z
I use the following function to convert it to HL7
`DateUtil.convertDate( "yyyy-MM-dd'T'HH:mm:ss", "yyyyMMddHHmmss", reportData.Documents.Pdf.CreatedInfo.LocalDate )`
And it works, the outcome is 20230601095139, like expected.

But if I include milliseconds, like the SimpleDateFormat implies
`DateUtil.convertDate( "yyyy-MM-dd'T'HH:mm:ss.SSS", "yyyyMMddHHmmssSSS", reportData.Documents.Pdf.CreatedInfo.LocalDate )`
I get this: 20230601122929000

I also tried adding the offset with yyyy-MM-dd'T'HH:mm:ss.SSSZ, yyyy-MM-dd'T'HH:mm:ss.SSSZZZZ, yyyy-MM-dd'T'HH:mm:ss.SSSX and yyyy-MM-dd'T'HH:mm:ss.SSSXXXX but I either get errors for not parseable dates or the same whacky result.

Any ideas here?

**Comment by jonbartels** [comment, marked as answer]
I tested the problem with this code. I get the same results

I get the same errors that you do:
```
var input = '2023-06-01T09:51:39.9470000Z';

var dates = {
	'input' : input,
	'noMilliSeconds': DateUtil.convertDate( "yyyy-MM-dd'T'HH:mm:ss", "yyyyMMddHHmmss", input ),
	'withMilliSeconds':DateUtil.convertDate( "yyyy-MM-dd'T'HH:mm:ss.SSS", "yyyyMMddHHmmssSSS", input ),
	//this errors 'withMilliSecondsAndZone' : DateUtil.convertDate( "yyyy-MM-dd'T'HH:mm:ss.SSSX", "yyyyMMddHHmmssSSS", input )
	//'withMilliSecondsExtraZeroes' : DateUtil.convertDate( "yyyy-MM-dd'T'HH:mm:ss.SSS0000", "yyyyMMddHHmmssSSS", input ),
	// errors 'withMilliSecondsExtraZeroesAndZone' : DateUtil.convertDate( "yyyy-MM-dd'T'HH:mm:ss.SSS0000Z", "yyyyMMddHHmmssSSS", input )
	'withNanoSeconds':DateUtil.convertDate( "yyyy-MM-dd'T'HH:mm:ss.SSSSSS", "yyyyMMddHHmmssSSS", input ),
	'withMilliSecondsLiteralZ' : DateUtil.convertDate( "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", "yyyyMMddHHmmssSSS", input ),
	'thisIsDumbButItWorks' : DateUtil.convertDate( "yyyy-MM-dd'T'HH:mm:ss.SSS", "yyyyMMddHHmmssSSS", input.substring(0, 23))
}

return JSON.stringify(dates);
```

Is the `0000` indicating nanoseconds or indicated the GMT time zone?

My `thisIsDumbButItWorks` entry trims the extra zeros and Z off of the input string then parses it. I do not like this strategy, it works for this test case but could fail in weird ways.

**Reply by NoraTheExplorer** [reply]
`thisIsDumbButItWorks ` wrong name! It's never dumb if it works 😄 
But you're right, it might fail on other occasions. 

It's just very weird and right now I don't know what the '0000Z' is supposed to mean. It could be an offset but none of the patterns seem to work. 

EDIT: The TIMESTAMP is not ISO, It's RFC 3339. Good to know.,
There are actaully workarounds for that: http://www.java2s.com/example/java-utility-method/date-rfc-format/get3339date-string-rfc3339date-2e13d.html

**Reply by tonygermano** [reply]
It looks like [java.text.SimpleDateFormat](https://docs.oracle.com/javase/8/docs/api/java/text/SimpleDateFormat.html) (used by mirth's DateUtil) does not support nanoseconds, but [java.time.format.DateTimeFormatter](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html) does.
